user module password expiration fixes

lilatomic opened the pull request:
##### SUMMARY
Fixes #75017

- can set password_expiry_min to 0 (any time)
- can set password_expire_max to 0 (always, seems mean)
bonus:
- can set both password_expiry_min *and* password_expiry_max at the same time
- checks the output of `chage` calls to see if the task failed


##### ISSUE TYPE
Bugfix Pull Request


##### COMPONENT NAME
ansible.builtin.user

##### ADDITIONAL INFORMATION

This time with the right issue number! so now the other issue doesn't have a reference to an irrelevant PR. Sorry again about that.

lilatomic committed:
allow inputting 0 for password_expire_{min|max}

0 is meaningful for min days (any time)
0 is technically valid for max_days

lilatomic committed:
[0] return result of execute_command from set_password_expire*

lilatomic committed:
[1] better return for set_password_expire

lilatomic committed:
add test for setting both min and max expiry

lilatomic committed:
[2] handle returns from set_password_expire*

lilatomic committed:
add change fragment

samdoran commented on lib/ansible/modules/user.py:
```diff
@@ -3195,14 +3193,28 @@ def main():
+    if user.password_expire_max is not None:
         if user.user_exists():
             (rc, out, err) = user.set_password_expire_max()
+            if rc is None:
+                pass  # target state reached, nothing to do
+            else:
+                if rc != 0:
+                    module.fail_json(name=user.name, msg=err, rc=rc)
+                else:
+                    result['changed'] = True
 
     # deal with password expire min
-    if user.password_expire_min:
+    if user.password_expire_min is not None:
         if user.user_exists():
             (rc, out, err) = user.set_password_expire_min()
+            if rc is None:
+                pass  # target state reached, nothing to do
+            else:
+                if rc != 0:
+                    module.fail_json(name=user.name, msg=err, rc=rc)
+                else:
+                    result['changed'] = True
```
This is doing the right thing but I think it can be consolidated a bit.

First, it should be able to be indented so it falls under the conditional from line 3156. I see no reason why these commands should be run if the user account does not exist or the state is set to `absent`.

Once indented, these two chunks of code can be streamlined.

```python
        # deal with password expiration
        for attr in 'min', 'max':
            if getattr(user, 'password_expire_%s' % attr) is not None:
                (rc, out, err) = getattr(user, 'set_password_expire_%s' % attr)()
                if rc is not None:
                    if rc != 0:
                        module.fail_json(name=user.name, msg=err, rc=rc)
                    else:
                        result['changed'] = True
```

lilatomic replied:
I agree that this should be moved into the conditional.

I'd like to find an alternative to the use of reflection here. It hides the references to the variables and functions, and makes finding their usage harder. For example, if you hadn't pointed out here that we shouldn't set the expiry if we remove a user, someone might have come in the future to fix that. Finding _where_ the functions are invoked could then not be done with a "find references" or "search" command.

I've proposed making an in-place helper function 6e6970ebdc2b0edfaaae07ff4d01b04e742e9493 , I think we get both the deduplication of code and the ease of reference for not many more lines.

samdoran replied:
I agree that hiding the reference to the variables is a bit opaque. I would rather avoid an internal function since they are harder to target for testing.

samdoran replied:
Perhaps setting the expiration needs to be in its own method since the output of setting both min and max should be aggregated. Something like this may work:

```patch
diff --git lib/ansible/modules/user.py lib/ansible/modules/user.py
index f8cd675bbf..775bc9a244 100644
--- lib/ansible/modules/user.py
+++ lib/ansible/modules/user.py
@@ -1046,6 +1046,30 @@ class User(object):
         else:
             return self.execute_command(cmd)
 
+    def set_password_expiration(self):
+        rc, out, err = (None, '', '')
+
+        if self.password_expire_max is not None:
+            _rc, _out, _err = self.set_password_expire_max()
+            if _rc is not None:
+                rc = _rc
+
+            out += _out
+            err += _err
+
+        if self.password_expire_min is not None:
+            _rc, _out, _err = self.set_password_expire_min()
+            if _rc is not None:
+                rc = _rc
+
+            out += _out
+            err += _err
+
+        if rc not in [None, 0]:
+            self.module.fail_json(name=self.name, msg=err, rc=rc)
+
+        return rc, out, err
+
     def user_password(self):
         passwd = ''
         expires = ''
@@ -3192,20 +3216,9 @@ def main():
             result['ssh_key_file'] = user.get_ssh_key_path()
             result['ssh_public_key'] = user.get_ssh_public_key()
 
-        def _handle_set_password_expire_return(command_return):
-            (rc, out, err) = command_return
-            if rc is None:
-                pass  # target state reached, nothing to do
-            else:
-                if rc != 0:
-                    module.fail_json(name=user.name, msg=err, rc=rc)
-                else:
-                    result['changed'] = True
-
-        if user.password_expire_max is not None:
-            _handle_set_password_expire_return(user.set_password_expire_max())
-        if user.password_expire_min is not None:
-            _handle_set_password_expire_return(user.set_password_expire_min())
+        rc, out, err = user.set_password_expiration()
+        if rc is not None:
+            result['changed'] = True
 
     module.exit_json(**result)
 
```

samdoran replied:
There is room for improvement with that patch since the stdout from the mix/max commands isn't being returned on success. Maybe that's unnecessary?

lilatomic replied:
hm, it looks like I never hit submit on my comment:
It turns out that you can set both at the same time, so we don't need to aggregate the output. The current revision has some duplication and it _feels_ like we should be able to pull that out, but it ends up being more trouble than it's worth, IMO.

samdoran commented on changelogs/fragments/75017-user-password-expiry.yml:
```diff
@@ -0,0 +1,5 @@
+bugfixes:
+  - can set password_expiry_min to 0
+  - can set password_expiry_max to 0
+  - can set both password_expiry_min and password_expiry_max at the same time
+  - checks the output from `chage` calls to see if the task failed
```
```suggestion
  - user - allow ``password_expiry_min`` and ``password_expiry_min`` to be set to ``0`` (https://github.com/ansible/ansible/issues/75017)
  - user - allow password min and max to be set at the same time (https://github.com/ansible/ansible/issues/75017)
```

lilatomic committed:
only set password expiry if user exists

lilatomic committed:
remove now redundant condition

already asserted above

lilatomic committed:
collect return-handling code

lilatomic committed:
concise doc fragment

Co-authored-by: Sam Doran <sdoran@redhat.com>

lilatomic commented:
@samdoran , I think I've addressed your feedback, would you be able to re-review this?

lilatomic committed:
combine password min and max into one execution

lilatomic committed:
handle case where spwd is not present

like on macOS and FreeBSD

lilatomic committed:
fix tests

I think this is a result of the big revamp of testing

Sometimes running the tests as a whole suite causes this to work,
since one of the earlier modules creates the folder.
But that means that depending on which modules run (and which clean up!)
results will vary
So hopefully this ensures it is always present

mkrizek commented:
!waffling core_review

mkrizek commented:
!waffling needs_revision
